Create a C++ sample for AB-testing plugin.

amalykh opened the pull request:
Fixes #<issue_number_goes_here>

> It's a good idea to open an issue first for discussion.

- [ ] Tests pass
- [ ] Appropriate changes to documentation are included in the PR

amalykh commented:
/assign @nkllkc

leonm1 commented on samples/ab_testing/test.cc:
```diff
@@ -0,0 +1,64 @@
+      EXPECT_THAT(res.headers, ElementsAre(Pair(":path", "example.com/file2.blah")));
+      file2_served = true;
+    }
+  }
```
Nit: lines 41 and 46 are two test cases deserving their own TEST_P(HttpTest, UnrelatedPathUnchanged) and TEST_P(HttpTest, ExperimentPathExercised)

amalykh commented:
Hi [leonm1](https://github.com/leonm1), I have resolved the outstanding comments. Please take a look and see if we can merge the PR in.

martijneken commented on samples/ab_testing/plugin.cc:
```diff
@@ -0,0 +1,46 @@
+
+  FilterHeadersStatus onRequestHeaders(uint32_t headers,
+                                       bool end_of_stream) override {
+    std::string path = getRequestHeader(":path")->toString();
```
toString() forces a string copy, whereas WasmDataPtr->view() will return std::string_view to avoid copies.

You may also want an existence check on the :path header, although to be fair it should always exist.

amalykh replied:
Great point, I have added a check for the path (just in case) and changed this line to use string view.

martijneken commented on samples/ab_testing/plugin.cc:
```diff
@@ -0,0 +1,46 @@
+    const std::string experiment_file = "/file2.blah";
+    if (path.length() >= default_file.length() &&
+        std::equal(default_file.rbegin(), default_file.rend(), path.rbegin())) {
+      bool serve_alternative_file = (rand() % 2 == 1);
```
A little more challenging, but I wonder if we could demonstrate how to implement arbitrarily weighted choices, e.g. 80% A, 20% B.

amalykh replied:
Agree, this sounds like a useful use-case to show. I have modified the code correspondingly.

martijneken commented on samples/ab_testing/plugin.cc:
```diff
@@ -0,0 +1,46 @@
+    const std::string default_file = "/file1.blah";
+    const std::string experiment_file = "/file2.blah";
+    if (path.length() >= default_file.length() &&
+        std::equal(default_file.rbegin(), default_file.rend(), path.rbegin())) {
```
I got a little lost in this logic -- is it a substring check on the full path? Worth a comment.,

You might also consider using URL parsing (and manipulation below) from boost, which I am adding in PR #17.

amalykh replied:
I have mode some refactoring here by moving this if condition to a separate function and adding around some comments. Please, take a look if it's more readable or whether you would like me to improve it further.

As for using the Boost library: I have a mild preference to have as few dependecies as we can get. I think it's more beneficial to focus on a specific thing that we are trying to showcase with this sample.

martijneken commented:
It's worth adding a link to this sample to the README (maybe once you have the Rust sample ready)

amalykh commented:
Hi [martijneken](https://github.com/martijneken), thanks for your review and sorry for not noticing that it has happened for a long time. I have addressed the comments and I would appreciate if you could make another pass.

As for the README file -- this should be straighforward, I'm happy to do it as part of the Rust PR.

martijneken reviewed (approved):
LGTM, thanks!

amalykh commented on samples/ab_testing/plugin.cc:
```diff
@@ -15,25 +15,45 @@
+    if (!getRequestHeader(":path")) {
+      return FilterHeadersStatus::Continue;
+    }
+    std::string_view path = getRequestHeader(":path")->view();
```
Good point, done.

msampathkumar commented:
@amalykh - This PR seems ready, PTAL.
